src/main/scala/gen/gmos/GmosNLongslitD.scala

```diff
@@ -326,21 +322,39 @@ object GmosNLongslitD {
       // Computes the science sequence as a stream of "science/flat" "atoms"
       // where the offset in Q and observing wavelength vary. Continues until
       // the provided `reachedS2N` effect evalues `true`.
-      def science(
-        itc:        ImaginaryItc[F],
-        reachedS2N: F[Boolean]
-      ): Stream[F, Stream[F, Step.GmosN]] = {
+      override def science(
+        reachedS2N: Int => F[Boolean]
+      ): Stream[F, Stream[F, (Step.GmosN, Double)]] = {
+
+        val steps = ScienceSteps(observingMode)
 
-        val steps = ScienceSteps(mode)
+        // Updates science step exposure time and adds this step's contribution
+        // to total s/n.
+        def applyItc(step: Step.GmosN, s: Itc.Result.Success): (Step.GmosN, Double) =
+          step match {
+
+            case Step.GmosN(_, Science(_)) =>
+              (exposureTime.optional.set(s.exposureTime)(step), s.stepSignalToNoise)
+
+            case _                         =>
+              (step, 0.0)
+          }
 
         // A mini stream containing two steps, one a science dataset and the
         // other a smart flat.  These shouldn't be broken up by acquisition so
         // we'll make the nested stream visible to the caller.
-        def substream(a: Step.GmosN, b: Step.GmosN): Stream[F, Step.GmosN] =
-          Stream.force(itc.integrationTime(mode, magnitude).map { fd =>
-            val f = exposureTime.optional.set(fd)
-            Stream(f(a), f(b)).covary[F]
-          })
+        def substream(a: Step.GmosN, b: Step.GmosN): Stream[F, (Step.GmosN, Double)] =
+          Stream.force {
+            itc.calculate(targetProfile, observingMode, signalToNoise).map {
```

Comment:
Here we'll eventually need to factor in the current conditions.

Reply:
Yes, I'll externalize that soon. We can define an interface like `Conditions[F]` that we can stub out initially to simulate changing conditions. There is also the chance that ITC will say you can't continue in current conditions, so we'll need to change ITC and this code handle that case.